### PR TITLE
Add Autofill functionality

### PR DESCRIPTION
**Pull request by maxibor**
Given a ENA accession (project or sample), uses the ENA Portal API to autofill a libraries metadata sample

```bash
$ AMDirT autofill --help
Usage: AMDirT autofill [OPTIONS] ACCESSION OUTPUT

  Autofills library table from using ENA accession numbers

  ACCESSION: ENA accession number
  OUTPUT: path to output table file

Options:
  -t, --table_name [ancientmetagenome-environmental|ancientmetagenome-hostassociated|ancientsinglegenome-hostassociated|test]
                                  [default: ancientmetagenome-hostassociated]
  -a, --accession_type [project|sample]
                                  [default: sample]
  --help                          Show this message and exit.
```

```bash
$ AMDirT autofill -t test SAMN17080920 test2.tsv
valid_libaries.tsv is valid
2023-03-06 15:47:51.273 INFO    AMDirT: valid_libaries.tsv is valid
ENA API is up
2023-03-06 15:47:51.461 INFO    AMDirT: ENA API is up
Found 1 libraries
2023-03-06 15:47:54.826 INFO    AMDirT: Found 1 libraries
Writing libraries metadata to test2.tsv
2023-03-06 15:47:54.826 INFO    AMDirT: Writing libraries metadata to test2.tsv
$ cat test2.tsv
project_name	publication_year	data_publication_doi	sample_name	archive	archive_project	archive_sample_accession	library_name	strand_type	library_polymerase	library_treatment	library_concentration	instrument_model	library_layout	library_strategy	read_count	archive_data_accession	download_links	download_md5s	download_sizes
			11105_AfrSudan1		PRJNA685265	SAMN17080920	11105_AfrSudan1					Illumina HiSeq 2500	PAIRED	WGS	12036888	SRR13263116	ftp.sra.ebi.ac.uk/vol1/fastq/SRR132/016/SRR13263116/SRR13263116_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR132/016/SRR13263116/SRR13263116_2.fastq.gz	be49abcb2355e9c53ad450a8710876bc;3b008f806a85dc2f14b33ee27601a189	413063178;475748313
```

Limitations for now: 
- one accession at a time
- logging needs to be improved (double logging somehow)
- Won't work against AncientMetagenomeDir master branch, because schema on master  are not valid as of now (need to remove multiple types of download_sizes in schema)

### CODE REVIEW

**Review by jfy133 — CHANGES_REQUESTED** [review]
I've updated the `master` of the Dir to ensure single-data type columns (edit: properly now), so you should be able to test on the real data.

Overall it seems to work with no execution errors, however some feedback and two bugs in the output:

I can confirm I also get the double logging.

1. The sample name is wrong in a lot of cases (we've had to manually curate this in many cases), so I would recommend not including this (as people may not check this properly)
    - Technically this would also apply to platform, strategy, but these are at least 'fixed' lists and we can emphasise checking these as they are 'new' columns that don't have precendent in the sample table, so it's less problematic if we get this wrong in the initial library submission.
2. The `archive_sample_accession` has picked the primary, not the secondary accession (i.e., it seems to report the SAMN rather than SRS number) [see below [^1] for command and data]
3. ~~Now I've fixed `master` to have single-type columns, we lose the `download_size` information (as now it would be spread across `download_size_{1,2,3}`, so this will need ot be updated now.~~ EDIT: Not split anymore, works as expected

Finally: could the output file be specified by a flag `-o library_output.tsv`, and then a user can supply multiple accession codes as positional arguments, and then `autofill` loops through each API call internally?

e.g.

```
AMDirT autofill -t ancientmetagenome-hostassociated -a sample -o library_output.tsv ERS01234 ERS012345 ERS01267
```

[^1]: test commands: ```AMDirT autofill -t ancientmetagenome-hostassociated -a sample SRS473742 test.tsv``` and data  [test.zip](https://github.com/SPAAM-community/AMDirT/files/10946086/test.zip)

**Comment by maxibor** [comment]
Point 2 and final point are addressed.
I would keep it as is for 1. and add a warning

File: AMDirT/autofill/__init__.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    libraries = DatasetValidator(
+        schema=libraries_schema[table_name], dataset=libraries[table_name]
+    )
+
+    libraries.to_rich()
+
+    sample_df = sample.dataset.iloc[:0, :].copy()
+    libraries_df = libraries.dataset.iloc[:0, :].copy()
+
+    print(sample_df.columns)
```
Comment (jfy133) [inline AMDirT/autofill/__init__.py:52]:
```suggestion
```

**Review by jfy133 — APPROVED** [review]
Almost there I think!

I think in terms of supplying the actual accessions themselves, the positional arguments were OK before, as it would be easy to supply a list and print it in space separated list using a bash function/redirect. But conceptually what you have now 'works'.

But the specific flags for `-s` and `-l` are good.

For the warning about the sample name what about something:

```
WARNING: sample name must match that reported in publication and/or sample-level table. ENA reported sample-name may be correct! Check before submission.
```

We could save the stdout to a file to print as text in the GHA comment with the file (like we already do for AncientMetagenomeDirCheck during validation)?

I noticed a left-over debugging print statement (see suggestion), and I note we are still getting double logging, but I don't think that is critical

**Comment by maxibor** [comment]
I added a merge command to merge new dataset to the existing ones.
The command performs a few basic checks (schema, column duplication and line duplication) before merging the new dataset to the remote one.
As a bonus, there is now unduplicated  colored logging  :)

CLI:
```bash
$ AMDirT merge --help
Usage: AMDirT merge [OPTIONS] DATASET

  Merges new dataset with existing table


  DATASET: path to tsv file of new dataset to merge

Options:
  -n, --table_name [ancientmetagenome-environmental|ancientmetagenome-hostassociated|ancientsinglegenome-hostassociated|test]
                                  [default: ancientmetagenome-hostassociated]
  -t, --table_type [samples|libraries]
                                  [default: libraries]
  -m, --markdown                  Output is in markdown format
  -o, --outdir PATH               path to sample output table file  [default:
                                  .]
  --help                          Show this message and exit.
$ AMDirT merge -t samples -n ancientsinglegenome-hostassociated ancientsinglegenome-hostassociated_new_data.tsv
AMDirT [INFO]: New Dataset is valid
AMDirT [INFO]: Merging new dataset with remote ancientsinglegenome-hostassociated samples dataset
AMDirT [INFO]: New ancientsinglegenome-hostassociated samples dataset written to ./ancientsinglegenome-hostassociated_samples.tsv
```

File: AMDirT/autofill/__init__.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import json
+
+import sys
+import pandas as pd
+
+def run_autofill(accession, table_name=None, schema=None, dataset=None, sample_output=None, library_output=None, verbose=False):
+    """Autofill the metadata of a table from ENA
+
+    Args:
+        accession (str): ENA project accession. Multiple accessions can be space separated (e.g. PRJNA123 PRJNA456)
```
Comment (jfy133) [inline AMDirT/autofill/__init__.py:14]:
```suggestion
        accession (str): ENA project accession. Multiple accessions can be quote-wrapped and space separated (e.g. 'PRJNA123 PRJNA456')
```

**Review by jfy133 — APPROVED** [review]
Working for me other than the one change to the help

**Pull request merged**